### PR TITLE
Initial commit for UART API

### PR DESCRIPTION
**Pull request by Jeremy-Chau**

### CODE REVIEW

File: firmware/library/L1_Drivers/uart.hpp [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#pragma once
+
+class UARTInterface
+{
+ public:
+    virtual void setBaud(unsigned int baudrate) = 0;
```
Comment (kammce) [inline firmware/library/L1_Drivers/uart.hpp]: Follow Google C++ style guide. Should be `SetBaud`.
`int` has different size depending on what system you are compiling for, please use <cstdint>

File: firmware/library/L1_Drivers/uart.hpp [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+    virtual bool init(unsigned int baud, uint32_t queue_size = 16) = 0;
+    virtual void IntHandle() = 0;
```
Comment (kammce) [inline firmware/library/L1_Drivers/uart.hpp]: Write the full name of the function and what it is doing. A function or method should say and action and not be a nown. For example, would prefer `AttachInterruptHandler(...)`

Also, the method should accept a function pointer.

File: firmware/library/L1_Drivers/uart.hpp [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#pragma once
+
+class UARTInterface
+{
+ public:
+    virtual void setBaud(unsigned int baudrate) = 0;
+    virtual bool init(unsigned int baud, uint32_t queue_size = 16) = 0;
```
Comment (kammce) [inline firmware/library/L1_Drivers/uart.hpp]: Prefer to spell words out fully, for example `Initialize`

File: firmware/library/L1_Drivers/uart.hpp [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    virtual void IntHandle() = 0;
+    virtual void send(char out, unsigned int time_limit) = 0;
```
Comment (kammce) [inline firmware/library/L1_Drivers/uart.hpp]: Ditto from the above comment about stdint and naming.

File: firmware/library/L1_Drivers/uart.hpp [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#pragma once
+
+class UARTInterface
```
Comment (kammce) [inline firmware/library/L1_Drivers/uart.hpp]: See the naming rules for classes here: https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

**Comment by kammce** [comment]
Please squash your commits into 1.

File: firmware/library/L1_Drivers/uart.hpp [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <cstdint>
+
+class UART_Interface
```
Comment (kammce) [inline firmware/library/L1_Drivers/uart.hpp]: No underscore.

File: firmware/library/L1_Drivers/uart.hpp [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+{
+ public:
+    virtual void SetBaud(uint32_t baudrate) = 0;
+    virtual bool Initialize(uint32_t baud, uint32_t queue_size = 16) = 0;
+    virtual void InteruptHandler(void) = 0;
```
Comment (kammce) [inline firmware/library/L1_Drivers/uart.hpp]: What does this do? I still can't figure out what  UART's would do with this implemented method.

File: firmware/library/L1_Drivers/uart.hpp [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <cstdint>
+
+class UART_Interface
+{
+ public:
+    virtual void SetBaud(uint32_t baudrate) = 0;
+    virtual bool Initialize(uint32_t baud, uint32_t queue_size = 16) = 0;
```
Comment (kammce) [inline firmware/library/L1_Drivers/uart.hpp]: You should also provide methods for setting queue_size.

File: firmware/library/L1_Drivers/uart.hpp [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    virtual void InteruptHandler(void) = 0;
+    virtual void SendData(char out, uint32_t time_limit) = 0;
+    virtual bool ReceiveData(char *CharInput, uint32_t time_limit) = 0;
```
Comment (kammce) [inline firmware/library/L1_Drivers/uart.hpp]: variables (which includes parametes) should be snake_case.

Comment (kammce) [reply firmware/library/L1_Drivers/uart.hpp]: Prefer just the word `Send` and just `Recieve`

File: firmware/library/L1_Drivers/uart.hpp (hunk 7 above)
Comment (Jeremy-Chau) [reply firmware/library/L1_Drivers/uart.hpp]: This method would be called whenever an interrupt is created due to the Rx pin receiving data. The method would take that data from the pin and throw it onto the queue waiting to be picked up by the user. The user would then use the receive function to grab that data whenever they need to.

File: firmware/library/L1_Drivers/uart.hpp (hunk 8 above)
Comment (Jeremy-Chau) [reply firmware/library/L1_Drivers/uart.hpp]: You stated during the design doc phase that you didn't want the queue_size to be dynamically set. This is why I didn't provide a method to set the queue size before. Did you want to have queue_size dynamically set now?

Comment (kammce) [reply firmware/library/L1_Drivers/uart.hpp]: My original thought was that you were creating a queue, but now I see this refers to the FIFO (correct me if I'm wrong). Also, looking at the datasheet, I don't see a register that allows you to change the FIFO length. If that is the case, then remove queue_size as an option.

File: firmware/library/L1_Drivers/uart.hpp (hunk 7 above)
Comment (kammce) [reply firmware/library/L1_Drivers/uart.hpp]: I dont thing that the interface should include an InterruptHandler. You cannot attach an interrupt to a class's method unless it is static. Therefor, there would need to be a templated static handler for the various UART types. So this method should be removed.

File: firmware/library/L1_Drivers/uart.hpp (hunk 8 above)
Comment (Jeremy-Chau) [reply firmware/library/L1_Drivers/uart.hpp]: What if the queue_size parameter is used for a FreeRTOS queue being initialized in the function?

**Pull request merged**